### PR TITLE
Flexibility changes

### PR DESCRIPTION
**Pull request by kim-dongit**
Some changes to improve on flexibility. The documentation will need updating because of this, especially because the migrations are now copied by the publish command. This is a breaking change. On the other hand, it grants users the control they need to adjust to their application.

### CODE REVIEW

File: src/config/twofactor-auth.php [hunk 1]
```diff
@@ -2,6 +2,21 @@
...
+    |
+    | Options:
+    | - true: always require two-factor authentication
+    | - false: never require two-factor authentication
+    | - 'per_user': look if a row exists in the two_factor_auths table for the
+    |   user
+    |
+    */
+
+    'enabled' => 'per_user',
```
Comment (michaeldzjap) [inline src/config/twofactor-auth.php]: I don't really like the idea of mixing strings and booleans here. I'd rather see the three options: "always", "never", "user" or something.

**Comment by kim-dongit** [comment]
You are right. Made the necessary changes. Also added timestamps to the migration files that get published.

**Comment by michaeldzjap** [comment]
There's a couple of other things I'd like to see differently. Mostly style related, nothing major. Will comment on those now. But thanks for the PR. Useful additions.

File: src/Providers/BaseProvider.php [hunk 2]
```diff
@@ -18,6 +18,6 @@ public function enabled(User $user)
         if ($conf === 'per_user') {
             return !is_null($user->twoFactorAuth);
         }
-        return (bool) $conf;
+        return $conf === 'always';
```
Comment (michaeldzjap) [inline src/Providers/BaseProvider.php]: Please add a space between the closing `}` of the `if` and the `return` statement

File: src/TwoFactorAuthenticable.php [hunk 3]
```diff
@@ -31,7 +31,7 @@ public function setTwoFactorAuthId(string $id) : void
             // respect when 2fa is not set for user, never insert
             $this->twoFactorAuth->update(['id' => $id]);
         }
-        elseif ($enabled) {
+        elseif ($enabled === 'always') {
```
Comment (michaeldzjap) [inline src/TwoFactorAuthenticable.php]: I would prefer to change this to:

```php
$enabled = config('twofactor-auth.enabled', 'per_user');

if ($enabled === 'per_user') {
    $this->twoFactorAuth->update(['id' => $id]);
}

if ($enabled === 'always') {
    $this->upsertTwoFactorAuthId($id);
}

File: src/config/twofactor-auth.php [hunk 4]
```diff
@@ -8,8 +8,8 @@
     |--------------------------------------------------------------------------
     |
     | Options:
-    | - true: always require two-factor authentication
-    | - false: never require two-factor authentication
+    | - 'always': always require two-factor authentication
```
Comment (michaeldzjap) [inline src/config/twofactor-auth.php]: I would either change the options to: "always" and "never" or "enabled" and "disabled". Whether you go for the first or second I don't really care about.

The options "per_user" I would prefer to see as just "user".

File: src/TwoFactorAuthServiceProvider.php [hunk 5]
```diff
@@ -62,4 +60,21 @@ public function register()
...
+
+    /**
+     * Adds current timestamp prefix.
+     */
+    protected function publishMigrations()
+    {
+        $files = [
+            'add_mobile_to_users_table.php',
+            'create_two_factor_auths_table.php',
+        ];
```
Comment (michaeldzjap) [inline src/TwoFactorAuthServiceProvider.php:72]: Add a space between `]:` and `$paths`

File: src/TwoFactorAuthServiceProvider.php [hunk 6]
```diff
@@ -62,4 +60,21 @@ public function register()
...
+        $paths = [];
```
Comment (michaeldzjap) [inline src/TwoFactorAuthServiceProvider.php:74]: Add a space between this line and `foreach`

File: src/TwoFactorAuthServiceProvider.php [hunk 7]
```diff
@@ -62,4 +60,21 @@ public function register()
...
+        foreach ($files as $file) {
+            $paths[__DIR__ . '/database/migrations/' . $file] = database_path('migrations/'.date('Y_m_d_His').'_'.$file);
+        }
```
Comment (michaeldzjap) [inline src/TwoFactorAuthServiceProvider.php:78]: Add a space between the closing `}` of the `foreach` and the next line.

File: src/TwoFactorAuthServiceProvider.php [hunk 8]
```diff
@@ -62,4 +60,21 @@ public function register()
...
+     */
+    protected function publishMigrations()
+    {
+        $files = [
+            'add_mobile_to_users_table.php',
+            'create_two_factor_auths_table.php',
+        ];
+        $paths = [];
+        foreach ($files as $file) {
+            $paths[__DIR__ . '/database/migrations/' . $file] = database_path('migrations/'.date('Y_m_d_His').'_'.$file);
```
Comment (michaeldzjap) [inline src/TwoFactorAuthServiceProvider.php:77]: Don't use spaces between each `.`, so make it: `$paths[__DIR__.'/database/migrations/'.$file]`

File: src/TwoFactorAuthServiceProvider.php [hunk 9]
```diff
@@ -62,4 +60,21 @@ public function register()
             return $app->make(TwoFactorAuthManager::class)->provider();
         });
     }
+
+    /**
+     * Adds current timestamp prefix.
```
Comment (michaeldzjap) [inline src/TwoFactorAuthServiceProvider.php:65]: I'd like to see the following docstring:

```php
/**
 * Publish all migration files and prefix them with the current timestamp.
 *
 * @return void
 */
```

File: src/config/twofactor-auth.php [hunk 10]
```diff
@@ -64,4 +64,18 @@
...
+    |--------------------------------------------------------------------------
+    | Model setttings
+    |--------------------------------------------------------------------------
+    |
+    | \App\User is used, but its primary key can be specified
+    |
+    */
+    'models' => [
+        'user' => [
+            'pk' => 'id'
```
Comment (michaeldzjap) [inline src/config/twofactor-auth.php]: I'd like to have this renamed to "primaryKey" instead of "pk".

File: src/TwoFactorAuthenticable.php [hunk 11]
```diff
@@ -14,7 +14,7 @@ trait TwoFactorAuthenticable
      */
     public function twoFactorAuth() : HasOne
     {
-        return $this->hasOne(TwoFactorAuth::class);
+        return $this->hasOne(TwoFactorAuth::class, $this->getKeyName());
```
Comment (michaeldzjap) [inline src/TwoFactorAuthenticable.php]: I don't think this is correct. Shouldn't this be:

```php
return $this->hasOne(TwoFactorAuth::class, 'user_id', $this->getKeyName());
```

I.e. the second argument is the foreign key (i.e. "user_id" on the `two_factor_auths` table), the third argument the (custom) primary key.

File: src/TwoFactorAuth.php [hunk 12]
```diff
@@ -39,6 +39,6 @@ class TwoFactorAuth extends Model
      */
     public function user() : BelongsTo
     {
-        return $this->belongsTo(\App\User::class);
+        return $this->belongsTo(\App\User::class, config('twofactor-auth.models.user.pk', 'id'));
```
Comment (michaeldzjap) [inline src/TwoFactorAuth.php]: I don't think this is correct. Shouldn't this be:

```php
return $this->belongsTo(\App\User::class, 'user_id', config('twofactor-auth.models.user.primaryKey', 'id'));
```

I.e. the second argument is the foreign key (i.e. "user_id" on the two_factor_auths table), the third argument the (custom) primary key.

File: src/TwoFactorAuthenticable.php [hunk 13]
```diff
@@ -37,4 +45,19 @@ public function getTwoFactorAuthId() : string
     {
         return $this->twoFactorAuth->id;
     }
+
+    /**
+     * @param string $id
```
Comment (michaeldzjap) [inline src/TwoFactorAuthenticable.php:53]: I'd like to change the docstring to:

```php
/**
 * Create or update a two-factor authentication record with the given id.
 *
 * @param  string  $id
 * @return void
 */

File: src/TwoFactorAuthenticable.php [hunk 14]
```diff
@@ -37,4 +45,19 @@ public function getTwoFactorAuthId() : string
...
+     */
+    private function upsertTwoFactorAuthId(string $id) : void
+    {
+        DB::transaction(function () use ($id) {
+            $attributes = ['id' => $id];
+            if (!$this->twoFactorAuth()->exists()) {
```
Comment (michaeldzjap) [inline src/TwoFactorAuthenticable.php:61]: Insert a space between the assignment to `$attributes` and the `if` statement.

**Review by michaeldzjap — CHANGES_REQUESTED** [review]

**Comment by michaeldzjap** [comment]
Also, it would be good if you could add appropriate tests. There is a separate [test project](https://github.com/michaeldzjap/laravel-two-factor-authentication-example) for this package.

With regards to the breaking changes due to the publishing of the migrations files; this needs to be documented in the README. One problem I foresee is when you already have ran migrations for this package for the main database (and hence, don't want to run them again), but want to run them every time for your test database. I guess wrapping the `Schema::create()` calls inside an `if (environment('test')) {}` check would mitigate this problem.

File: src/TwoFactorAuthenticable.php (hunk 11 above)
Comment (kim-dongit) [reply src/TwoFactorAuthenticable.php]: You are entirely right.

**Comment by michaeldzjap** [comment]
Just published [v2.4.0](https://github.com/michaeldzjap/laravel-two-factor-authentication/tree/v2.4.0) which includes your additions. Thanks again.

**Comment by kim-dongit** [comment]
Thank you too. I am not able to look at tests now, but maybe later. What tests do you envision?

**Comment by michaeldzjap** [comment]
Some unit tests can be added that test the extended "enabled" options. At the moment only the "user" case is tested.

And similarly, some browser (Dusk) tests should really be added that check the extended "enabled" logic works as expected for the "always" and "never" cases.

I don't mind doing this myself, but I don't have much time either at the moment.

**Pull request merged**